### PR TITLE
Test generator for output consistency checks (V2): aggregates

### PR DESCRIPTION
**Pull request by nrainer-materialize**
This closes https://github.com/MaterializeInc/materialize/issues/19439, which is part of https://github.com/MaterializeInc/materialize/issues/18858.

### Content

* support [vertical data storage for aggregations ("multi-row structure")](https://github.com/MaterializeInc/materialize/blob/057851684c4d7d1c4f115eeda084b1202f76d63a/test/output-consistency/README.md#value-storage-layouts)
* support aggregation functions
* specification of most aggregation functions
* improved ignore mechanism for known inconsistencies
* printer to generate and print a minimal reproducible code sample when an inconsistency is detected
* other improvements

### Identified issues
* https://github.com/MaterializeInc/materialize/issues/19511

### CODE REVIEW

**Comment by nrainer-materialize** [comment]
Nightly build: https://buildkite.com/materialize/nightlies/builds/2475

File: misc/python/materialize/output_consistency/ignore_filter/inconsistency_ignore_filter.py
Comment (def-) [inline misc/python/materialize/output_consistency/ignore_filter/inconsistency_ignore_filter.py:88]: could merge with the previous if condition

File: misc/python/materialize/output_consistency/input_data/test_input_data.py
Comment (def-) [inline misc/python/materialize/output_consistency/input_data/test_input_data.py:53]: Could do `max_value = max(max_value, value_count)`

File: misc/python/materialize/output_consistency/input_data/test_input_data.py
Comment (def-) [inline misc/python/materialize/output_consistency/input_data/test_input_data.py:50]: Something like this (untested):
```python
return max(len(type_with_values.raw_values) for type_with_values in self.all_data_types_with_values)
```

File: misc/python/materialize/output_consistency/output/format_constants.py
Comment (def-) [inline misc/python/materialize/output_consistency/output/format_constants.py:11]: Does this work with buildkite output? https://buildkite.com/docs/pipelines/managing-log-output

File: misc/python/materialize/output_consistency/ignore_filter/inconsistency_ignore_filter.py
Comment (def-) [inline misc/python/materialize/output_consistency/ignore_filter/inconsistency_ignore_filter.py:85]: Do we actually still generate those queries or just not generate them at all then? If we do, could be interesting to keep track of the number of queries failing with this issue.

**Comment by def-** [comment]
While running it locally I noticed that there are quite some errors. Can you print the number of queries with errors? Maybe some common patterns can be refined.
Examples:
```
-- * DB error in 'Dataflow rendering' was: function "array_arg" does not exist (No function matches the given name and argument types. You might need to add explicit type casts.)
-- * DB error in 'Dataflow rendering' was: operator does not exist: double precision & bigint (No operator matches the given name and argument types. You might need to add explicit type casts.)

```

**Comment by def-** [comment]
Does this mean that we should get an error but don't get it? Should this cause a failure?
```
-- Test with query #75 PASSED (result data matches).
-- Remarks:
-- * DB error was expected
-- * Query success with single column (Query successfully returning a value could be run with other queries.)
```

**Comment by def-** [comment]
By running with `bin/mzcompose --find output-consistency run default --max-iterations 1000000000 --max-runtime-in-sec 100000000 --verbose | tee out`, and checking the output with `grep "DB error in" out | sort | uniq -c | sort -nr | grep -v "^      [0-9]"` I see these happening often:
```
    216 -- * DB error in 'Dataflow rendering' was: function "array_arg" does not exist (No function matches the given name and argument types. You might need to add explicit type casts.)
    164 -- * DB error in 'Dataflow rendering' was: Evaluation error: division by zero
     90 -- * DB error in 'Dataflow rendering' was: function "variance_samp" does not exist (No function matches the given name and argument types. You might need to add explicit type casts.)
     85 -- * DB error in 'Dataflow rendering' was: function "variance_pop" does not exist (No function matches the given name and argument types. You might need to add explicit type casts.)
     50 -- * DB error in 'Dataflow rendering' was: Evaluation error: function ln is not defined for zero
     42 -- * DB error in 'Dataflow rendering' was: Evaluation error: function log10 is not defined for zero
     38 -- * DB error in 'Dataflow rendering' was: Evaluation error: value out of range: overflow
     36 -- * DB error in 'Dataflow rendering' was: Evaluation error: function log is not defined for zero
     36 -- * DB error in 'Dataflow rendering' was: Evaluation error: function acosh is defined for numbers greater than or equal to 1
     33 -- * DB error in 'Dataflow rendering' was: Evaluation error: function atanh is defined for numbers between -1 and 1 inclusive
     30 -- * DB error in 'Dataflow rendering' was: Evaluation error: function acos is defined for numbers between -1 and 1 inclusive
     29 -- * DB error in 'Dataflow rendering' was: Evaluation error: function ln is not defined for negative numbers
     26 -- * DB error in 'Dataflow rendering' was: Expected an expression, found right parenthesis
     23 -- * DB error in 'Dataflow rendering' was: Evaluation error: function asin is defined for numbers between -1 and 1 inclusive
     20 -- * DB error in 'Dataflow rendering' was: Evaluation error: function log10 is not defined for negative numbers
     20 -- * DB error in 'Dataflow rendering' was: Evaluation error: cannot take square root of a negative number
     19 -- * DB error in 'Dataflow rendering' was: Evaluation error: function log is not defined for negative numbers
     18 -- * DB error in 'Dataflow rendering' was: Evaluation error: numeric field overflow
     15 -- * DB error in 'Dataflow rendering' was: Evaluation error: real out of range
     13 -- * DB error in 'Dataflow rendering' was: operator does not exist: ~ numeric (No operator matches the given name and argument types. You might need to add explicit type casts.)
     11 -- * DB error in 'Dataflow rendering' was: Evaluation error: integer out of range
```
At least the "function/operator does not exist" sounds like it should be fixed, the "Expected an expression" too.

File: misc/python/materialize/output_consistency/ignore_filter/inconsistency_ignore_filter.py
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/ignore_filter/inconsistency_ignore_filter.py]: ✅  Merged outer if-conditions. I am aware that the inner ones could also be merged but prefer to keep those separate.

File: misc/python/materialize/output_consistency/input_data/test_input_data.py
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/input_data/test_input_data.py]: ✅

File: misc/python/materialize/output_consistency/input_data/test_input_data.py
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/input_data/test_input_data.py]: ✅ Thanks

File: misc/python/materialize/output_consistency/output/format_constants.py
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/output/format_constants.py:11]: Thanks for the hint, will check and adjust.

File: misc/python/materialize/output_consistency/ignore_filter/inconsistency_ignore_filter.py
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/ignore_filter/inconsistency_ignore_filter.py:85]: Filtering happens after generation, we can count them. Will do.

**Comment by nrainer-materialize** [comment]
> Does this mean that we should get an error but don't get it? Should this cause a failure?
> [...]
> Edit: I guess this means "if an error happens here, we expect and ignore it, but the query might also succeed"

Your guess is correct. Would it be more clear if I use the term "assumed" instead of "expected"?

> While running it locally I noticed that there are quite some errors. Can you print the number of queries with errors? 

Not sure if this makes sense. I consider most of these (if not all) invalid expressions generated by the generator. I will have a look at the two error types below but we probably have to live with most other errors. The effort to reliably, statically determine if an expression will overflow or leave a certain numeric range is not reasonable.

> At least the "function/operator does not exist" sounds like it should be fixed, the "Expected an expression" too.

I will have a look at those two types.

File: misc/python/materialize/output_consistency/generators/expression_generator.py
Comment (philip-stoev) [inline misc/python/materialize/output_consistency/generators/expression_generator.py:106]: Do you have an example of this situation?

**Review by philip-stoev — APPROVED** [review]
This is quite a lot of code and commits, so I defer to @def- on this PR. Going forward I hope that the basic structure will become stable enough so that PRs will get smaller over time.

File: misc/python/materialize/output_consistency/generators/expression_generator.py
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/generators/expression_generator.py:106]: Given the following setup:
```
DROP TABLE IF EXISTS t_dfr_vert;
CREATE TABLE t_dfr_vert (row_index INT, decimal_39_0_val DECIMAL(39));
INSERT INTO t_dfr_vert VALUES (1, 0::DECIMAL(39));
INSERT INTO t_dfr_vert VALUES (9, -999999999999999999999999999999999999999::DECIMAL(39));
-- ------------------------------
-- Setup for evaluation strategy 'Constant folding':
CREATE OR REPLACE VIEW v_ctf_vert (row_index, decimal_39_0_val)
 AS SELECT 1, 0::DECIMAL(39)
    UNION SELECT 9, -999999999999999999999999999999999999999::DECIMAL(39);
```
an example is:
```
-- Query using evaluation strategy 'Dataflow rendering':
SELECT
  ln(decimal_39_0_val)
FROM
  t_dfr_vert
WHERE row_index IN (9, 1)
ORDER BY row_index ASC;
-- ------------------------------
-- Query using evaluation strategy 'Constant folding':
SELECT
  ln(decimal_39_0_val)
FROM
  v_ctf_vert
WHERE row_index IN (9, 1)
ORDER BY row_index ASC;
```

The column `decimal_39_0_val` contains zero and a negative value:
* Data-flow rendering will fail with: `function ln is not defined for negative numbers`
* Constant folding will fail with: `function ln is not defined for zero`

I consider both error messages valid.

**Comment by nrainer-materialize** [comment]
> This is quite a lot of code and commits, so I defer to @def- on this PR. Going forward I hope that the basic structure will become stable enough so that PRs will get smaller over time.

True, this pull request allowing an additional underlying data layout was quite invasive. I also expect future PRs to become smaller.

File: misc/python/materialize/output_consistency/generators/expression_generator.py
Comment (philip-stoev) [reply misc/python/materialize/output_consistency/generators/expression_generator.py:106]: ok understood.

Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/generators/expression_generator.py:106]: However, this will be a problem when adding the top-k functionality... 😞

Comment (vmarcos) [reply misc/python/materialize/output_consistency/generators/expression_generator.py:106]: Yes, I think this is fine, since we cannot guarantee evaluation order under different plans.

File: misc/python/materialize/output_consistency/validation/validation_message.py
Comment (def-) [inline misc/python/materialize/output_consistency/validation/validation_message.py:101]: These assertions on the inputs smell like there should be two different methods to initialize it instead, so that the constraint is entirely expressed through the parameter types.

**Review by def- — APPROVED** [review]
Fine by me in general. Some more minor nitpicks.

File: misc/python/materialize/output_consistency/output/format_constants.py
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/output/format_constants.py:11]: Tried it out, they don't collide but also don't bring benefits. I think the issue is that the string is prefixed with the timestamp:

```
[2023-05-30T10:10:04Z] -- +++ Test summary
[2023-05-30T10:10:04Z] -- 32985/32985 queries passed in mode 'LIVE_DATABASE'.
```

Comment (def-) [reply misc/python/materialize/output_consistency/output/format_constants.py:11]: Is that the timestamp that buildkite implicitly adds? If so, it shouldn't matter. The `--` looks strange though.

File: misc/python/materialize/output_consistency/output/reproduction_code_printer.py
Comment (def-) [reply misc/python/materialize/output_consistency/output/reproduction_code_printer.py:126]: It works if you keep related code in the same file:
```py
class A:
    def __init__(self):
        print(B)

class B (A):
    pass

b = B()
```

File: misc/python/materialize/output_consistency/output/format_constants.py
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/output/format_constants.py]: You are right. ✅ Adjusted.

File: misc/python/materialize/output_consistency/ignore_filter/inconsistency_ignore_filter.py
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/ignore_filter/inconsistency_ignore_filter.py:85]: Not addressed for now (see previous comment), let me know if this is important to you.

File: misc/python/materialize/output_consistency/validation/validation_message.py
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/validation/validation_message.py]: I had a look at it and figured out that the "value" and "sql" properties are actually always set. Therefore, I changed the type from Optional and simplified the code. Thanks for the hint!

✅

File: misc/python/materialize/output_consistency/output/reproduction_code_printer.py
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/output/reproduction_code_printer.py:125]: Noted down. To be done afterwards in a separate PR to avoid conflicts with other changes on other branches.

**Comment by nrainer-materialize** [comment]
I will merge the current state when tests are green. If there is anything else you would like me to change, I am happy to address that in a follow-up PR.

**Pull request merged**
